nautobot/dcim/models/devices.py

```diff
@@ -1209,3 +1210,105 @@ def __str__(self):
         if self.alias:
             return self.alias
         return f"{self.platform} - {self.version}"
+
+
+#
+# Controller
+#
+
+
+@extras_features(
+    "custom_links",
+    "custom_validators",
+    "dynamic_groups",
+    "export_templates",
+    "graphql",
+    "locations",
+    "statuses",
+    "webhooks",
+)
+class Controller(PrimaryModel, ConfigContextModel):
+    """Controller model."""  # TODO: (whitej6) finialize description
+
+    name = models.CharField(max_length=100, unique=True)
+    status = StatusField(blank=False, null=False)
+    description = models.CharField(max_length=200, blank=True)
+    location = models.ForeignKey(
+        to="dcim.Location",
+        on_delete=models.PROTECT,
+        related_name="controllers",
+    )
+    platform = models.ForeignKey(
+        to="dcim.Platform",
+        on_delete=models.SET_NULL,
+        related_name="controllers",
+        blank=True,
+        null=True,
+    )
+    role = RoleField(blank=False, null=False)
+    tenant = models.ForeignKey(
+        to="tenancy.Tenant",
+        on_delete=models.PROTECT,
+        related_name="controllers",
+        blank=True,
+        null=True,
+    )
+    external_integration = models.ForeignKey(
+        to="extras.ExternalIntegration",
+        on_delete=models.PROTECT,
+        related_name="controllers",
+        blank=True,
+        null=True,
+    )
+    deployed_controller_type = models.ForeignKey(
+        to=ContentType,
+        on_delete=models.SET_NULL,
+        related_name="+",
+        blank=True,
+        null=True,
+    )
+    deployed_controller_id = models.UUIDField(blank=True, null=True)
+    deployed_controller = GenericForeignKey(ct_field="deployed_controller_type", fk_field="deployed_controller_id")
+    device_groups = models.ManyToManyField(
+        to="dcim.DeviceGroup", related_name="controllers", through="ControllerDeviceGroupAssociation"
+    )
+
+
+@extras_features(
+    "custom_links",
+    "custom_validators",
+    "dynamic_groups",
+    "export_templates",
+    "graphql",
+    "locations",
+    "statuses",
+    "webhooks",
+)
+class DeviceGroup(TreeModel, PrimaryModel, ConfigContextModel):
+    """Controller model."""  # TODO: (whitej6) finialize description
+
+    name = models.CharField(max_length=48)
+    devices = models.ManyToManyField(
+        to="dcim.Device",
+        related_name="device_groups",
+    )
+
+
+class ControllerDeviceGroupAssociation(BaseModel):
+    """Controller model."""  # TODO: (whitej6) finialize description
+
+    weight = models.PositiveIntegerField(default=1000)
```

Comment:
Is the idea that we'd have a `Meta.ordering` based on the weight? And just to be clear, is this for weighting devices among a controller, or controllers among a device, or both?

Reply:
Yes. Thought is predictive ordering from `Controller` to `DeviceGroup` but technically would handle the same in reverse

Reply:
If you have 5 top level groups assigned to a controller and you want predictable ordering of the groups returned in the query. E.g. would enable the following example to be a predictable order when you do controller.device_groups

- ControllerA 
    - Group XYZ
    - group 123
    - Another group